Fix OpenJ9 Java 8 signature test failures

JasonFengJ9 opened the pull request:
Fix OpenJ9 Java 8 signature test failures

Removed protect modifier from method getTarget();
Removed public modifier from asTypeUncached().

Close: #531 

Reviewer @pshipton 
FYI: @DanHeidinga @adamfarley 

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

pshipton commented:
@JasonFengJ9 did you also try the Java 9 signature test?

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java:
```diff
@@ -1299,6 +1295,10 @@ BoundMethodHandle rebind() {
 	}
 	
 /*[IF Sidecar19-SE-OpenJ9]*/
+	protected MethodHandle getTarget() {
```
Does this need to be protected or can it be package private?

JasonFengJ9 replied:
`protected` seems needed by `DelegatingMethodHandle`
```
abstract class java.lang.invoke.DelegatingMethodHandle extends java.lang.invoke.MethodHandle {
  protected abstract java.lang.invoke.MethodHandle getTarget();
}
```

DanHeidinga replied:
a subclass can increase the visibility of a method from package to protected.  As DelegatingMH isn't API, it's free to have a protected method.

JasonFengJ9 replied:
Agreed. `getTarget()` can be made to package private which can pass signature tests for `java.lang.invoke.MethodHandle`.

Looked again the initial compilation failure due to removal of this method as following:
```
/tmp/bld_369599/openjdk/jdk/src/java.base/share/classes/java/lang/invoke/MethodHandleImpl.java:510: error: method does not override or implement a method from a supertype
        @Override
```
But `MethodHandleImpl` doesn't inherit from `MethodHandle`. 
```
abstract class java.lang.invoke.MethodHandleImpl {
}
```
It is the inner classes of `MethodHandleImpl` that attempt to override `getTarget()`
```
Compiled from "MethodHandleImpl.java"
final class java.lang.invoke.MethodHandleImpl$AsVarargsCollector extends java.lang.invoke.DelegatingMethodHandle {
protected java.lang.invoke.MethodHandle getTarget();
}
....
final class java.lang.invoke.MethodHandleImpl$WrappedMember extends java.lang.invoke.DelegatingMethodHandle {
protected java.lang.invoke.MethodHandle getTarget();
}
```
I just realized that J9 has a stub class `DelegatingMethodHandle` which doesn't have `getTarget()`, and inherits it from `MethodHandle` instead.

JasonFengJ9 commented:
```
./jdk/bin/java -showversion -cp /bluebird/javatest/lib/jck/java9/jck-9-ea-tck-b68/JCK-runtime-9/classes javasoft.sqe.tests.api.signaturetest.SignatureTestBase -TestURL file:/bluebird/javatest/lib/jck/java9/jck-9-ea-tck-b68/JCK-runtime-9/tests/api/signaturetest/sigtest.html#java_base -TestedModule java.base
openjdk version "9-internal"
OpenJDK Runtime Environment (build 9-internal+0-adhoc.openj9.openjdk)
Eclipse OpenJ9 VM (build 2.9, JRE 9 Linux amd64-64 Compressed References 20171108_369613 (JIT enabled, AOT enabled)
OpenJ9   - 5d7b23b
OMR      - e1404d5
OpenJDK  - ae4d07b based on jdk-9+181)
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.sun.tdk.signaturetest.loaders.TigerRefgClassDescrLoader (file:/bluebird/javatest/lib/jck/java9/jck-9-ea-tck-b68/JCK-runtime-9/classes/) to field java.io.InputStream.MAX_SKIP_BUFFER_SIZE
WARNING: Please consider reporting this to the maintainers of com.sun.tdk.signaturetest.loaders.TigerRefgClassDescrLoader
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Unresolved parameter {java.util.Optional%0}
Unresolved parameter {java.util.Optional%0}
SignatureTest report
Base version: 9
Tested version:
Check mode: src [throws normalized]
Constant checking: on


Missing Methods
---------------

java.lang.String:   method public final java.util.stream.IntStream java.lang.String.chars()
java.lang.String:   method public final java.util.stream.IntStream java.lang.String.codePoints()
java.lang.StringBuffer:                 method public final java.util.stream.IntStream java.lang.StringBuffer.chars()
java.lang.StringBuffer:                 method public final java.util.stream.IntStream java.lang.StringBuffer.codePoints()
java.lang.StringBuilder:                method public final java.util.stream.IntStream java.lang.StringBuilder.chars()
java.lang.StringBuilder:                method public final java.util.stream.IntStream java.lang.StringBuilder.codePoints()
java.lang.invoke.MethodHandles:         method public static java.lang.invoke.MethodHandle java.lang.invoke.MethodHandles.dropArgumentsToMatch(java.lang.invoke.MethodHandle,int,java.util.List<java.lang.Class<?>>,int)
java.lang.invoke.VarHandle:             method public final boolean java.lang.invoke.VarHandle.isAccessModeSupported(java.lang.invoke.VarHandle$AccessMode)
java.lang.invoke.VarHandle:             method public final java.util.List<java.lang.Class<?>> java.lang.invoke.VarHandle.coordinateTypes()

Added Methods
-------------

java.lang.ClassLoader:                  method protected java.io.InputStream java.lang.ClassLoader.getResourceAsStream(java.lang.String,java.lang.String) throws java.io.IOException
java.lang.String:   method public java.util.stream.IntStream java.lang.CharSequence.chars()
java.lang.String:   method public java.util.stream.IntStream java.lang.CharSequence.codePoints()
java.lang.StringBuffer:                 method public java.util.stream.IntStream java.lang.CharSequence.chars()
java.lang.StringBuffer:                 method public java.util.stream.IntStream java.lang.CharSequence.codePoints()
java.lang.StringBuilder:                method public java.util.stream.IntStream java.lang.CharSequence.chars()
java.lang.StringBuilder:                method public java.util.stream.IntStream java.lang.CharSequence.codePoints()
java.lang.invoke.MethodHandle:          method protected java.lang.invoke.MethodHandle java.lang.invoke.MethodHandle.getTarget()
java.lang.invoke.MethodHandles:         method public final static java.lang.invoke.MethodHandle java.lang.invoke.MethodHandles.dropArgumentsToMatch(java.lang.invoke.MethodHandle,int,java.util.List<java.lang.Class<?>>,int)
java.lang.invoke.VarHandle:             method public boolean java.lang.invoke.VarHandle.isAccessModeSupported(java.lang.invoke.VarHandle$AccessMode)
java.lang.invoke.VarHandle:             method public java.util.List<java.lang.Class<?>> java.lang.invoke.VarHandle.coordinateTypes()
java.net.URLClassLoader:                method protected java.io.InputStream java.lang.ClassLoader.getResourceAsStream(java.lang.String,java.lang.String) throws java.io.IOException

Missed Annotations
------------------

java.lang.Class:    newInstance():anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Compiler: anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.System:   runFinalizersOnExit(boolean):anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="1.2")

Added Annotations
-----------------

java.lang.Class:    newInstance():anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="1.1")
java.lang.System:   runFinalizersOnExit(boolean):anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="1.2")

duplicate messages suppressed: 1

test: Failed. 26 errors
STATUS:Failed.test cases: 1; all failed; first test case failure: test
```

It appears this `getTarget()` can't be added for Java 9 either. Still investigating.

DanHeidinga commented:
@JasonFengJ9 If you confirm this passes the Java 8 signature test, then we should merge it as is.  The remaining Java 9 signature test issues can be handled in a separate PR.

DanHeidinga commented:
Jenkins test sanity

JasonFengJ9 commented:
The changeset has been updated.
Verified that Java 8 sr6 passed Java 8 signature tests and OpenJ9 passed Java 9 signature tests for j.l.i.MethodHandle (other failures are documented at https://github.com/eclipse/openj9/issues/548.
Thanks @DanHeidinga
